Create Custom Methods in Enumerable Module

anewman15 opened the pull request:
In this PR, I created my own versions of the following enumerable methods with a "my_" prefix:

- each
- each_with_index
- select
- all?
- any?
- none?
- count
- map
- inject

I also modified the my_map method to make use of blocks and procs

anewman15 committed:
Create two methods: my_each and my_each_with_index

anewman15 committed:
Create stickler, rubocop and Gemfiles

anewman15 committed:
Create my_select method, change return values of my_each and my_each_with_index

anewman15 committed:
Create my_select and my_any methods

anewman15 committed:
Create my_none and my_count methods

anewman15 committed:
Remove parameters from method definitions and use self

anewman15 committed:
Give variables more descriptive names

anewman15 committed:
Create my_inject and my_map_proc methods

anewman15 committed:
Create my_map_proc_block method and add block_given? condition

anewman15 committed:
Rewrite my_all? method

anewman15 committed:
Modify my_any?, my_none? and my_count methods

anewman15 committed:
Fix rubocop errors

anewman15 committed:
Modified my_inject method

anewman15 committed:
Re-wrote my_all? method

anewman15 committed:
Rewrote my_any, my_none and my_count methods

anewman15 committed:
Modify my_count method

anewman15 committed:
Fix rubocop errors

anewman15 committed:
Fix rubocop issue

lekegitrepo commented:
#### Hi @anewman15 :clap: :clap:
##### You have done a good job :+1: but you still need to fix or improve the following methods :point_down: 

**_1. my_each_with_index:_**
  `(1..5).my_each_with_index { |element, index| puts "#{element} => #{index}" }`

**_2. my_map:_**
  `(1..4).my_map { |n| n * n }`

**_3. my_inject:_**
  `(5..10).my_inject(1, :*)`

**_4. And also make the `multiply_els` method accessible for test_**

**_5. Add a descriptive README.md_**

**_6. The Pull Requests should be from the feature branch to the `develop` i.e it should be `develop from custom_enumerables` not `master from custom_enumerables`_**

**_fix these and submit for review._**

### _Reviewed by Temitayo Adeleke._

lekegitrepo reviewed (changes requested):
#### Status: Request changes

anewman15 committed:
Add readme file and app screenshot

anewman15 committed:
Update readme file

anewman15 commented:
Thanks, @lekegitrepo. I have made the changes

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+module Enumerable
+  def my_each
+    modified_object = []
```
my_each does not generate any object. Here you are creating an array object.

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+    else
+      to_enum
+    end
+  end
```
`my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
If you try to execute the following code, it will get executed without any error. But you my_each does not interact like that. The problem is length. Range does not have any length. Same problem when trying to work with hash.
```ruby
(1...4).each { |num| puts num }
```
I have modified your method into the following way and It worked properly. So you can take a look at the following code and replace it with your my_each
```ruby
def my_each
    arr = is_a?(Array) ? self : to_a
    i = 0
    if block_given?
      while i < arr.length
        yield arr[i]
        i += 1
      end
      self
    else
      to_enum
    end
end
```

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+  end
+
+  def my_each_with_index
+    modified_object = []
```
my_each_with_index does not create any object. Remove the array. It only yields the given block with the passed values.

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+    else
+      to_enum
+    end
+  end
```
`my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
Same problem as I described in the above comment related to my_each. Tried to execute the following code and faced error
```ruby
(1...3).my_each_with_index { |val, idx| puts "#{val}->#{idx}"}
# should output the following in the terminal
=begin
1->0
2->1
=end
```
```ruby
{a: 1, b: 2}.my_each_with_index { |val, idx| puts "#{val}->#{idx}"}
# should output the following in the terminal
=begin
[:a, 1]->0
[:b, 2]->1

=end
```
Checkout the solution I have given in my_each. The solution for this scenario is in there.

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+
+  def multiply_els(array)
+    array.my_inject(:*)
+  end
```
multiply_els is not an enumerable method. So it should be placed outside of the module.

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+    else
+      to_enum
+    end
+  end
```
`my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
It can also perform with range. For example
```ruby
(1...3).my_select { |num| num > 0 } # should return [1, 2 ]
```
But yours throw an error.

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+    end
+  end
+
+  def my_all?(pattern = nil)
+    true if length.zero?
+    selected_array = []
+
+    i = 0
+    if block_given?
+      selected_array = my_select { |element| yield element }
+    elsif !block_given? && pattern
+      while i < length
+        selected_array << self[i] if pattern === self[i]
+        i += 1
+      end
+    elsif !block_given? && !pattern
+      while i < length
+        selected_array << self[i] if self[i]
+        i += 1
+      end
+    end
+    selected_array.length == length
+  end
```
`my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
my_all? can also interact with range. But your one is giving me an error. Please run the following code.
```ruby
(1...3).my_all? # should return true
(1...3).my_all?(Integer) # should return true
```
So all the operation it can perform with array should be able to do with range. The answer is in the same solution that I have given.

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+    selected_array.length == length
+  end
+
+  def my_any?(pattern = nil)
+    false if length.zero?
+    selected_array = []
+
+    i = 0
+    if block_given?
+      selected_array = my_select { |element| yield element }
+    elsif !block_given? && pattern
+      while i < length
+        selected_array << self[i] if pattern === self[i]
+        i += 1
+      end
+    elsif !block_given? && !pattern
+      while i < length
+        selected_array << self[i] if self[i]
+        i += 1
+      end
+    end
+    selected_array.length.positive?
+  end
```
`my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
It has the ability to perform with range.
```ruby
(1...3).my_any? # should return true
(1...3).my_any?(Integer) # should return true
```

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+    selected_array.length.positive?
+  end
+
+  def my_none?(pattern = nil)
+    true if length.zero?
+    selected_array = []
+
+    i = 0
+    if block_given?
+      selected_array = my_select { |element| yield element }
+    elsif !block_given? && pattern
+      while i < length
+        selected_array << self[i] if pattern === self[i]
+        i += 1
+      end
+    elsif !block_given? && !pattern
+      while i < length
+        selected_array << self[i] if self[i]
+        i += 1
+      end
+    end
+    selected_array.length.zero?
+  end
```
`my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
Same as above it can perform with range too.
```ruby
(1...3).my_none? # should return false
(1...3).my_any?(Integer) # should return false
```

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+  end
+
+  def my_count(*args)
+    count_array = []
```
I have never seen any example of count which generated an array. Can you provide me an example that supports your implementation? All I know is, it returns numeric numbers. Please remove the array declaration.

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+    selected_array.length.zero?
+  end
+
+  def my_count(*args)
+    count_array = []
+
+    i = 0
+    if block_given?
+      while i < length
+        count_array << self[i] if yield self[i]
+        i += 1
+      end
+    elsif !block_given? && !args.empty?
+      while i < length
+        count_array << self[i] if self[i] == args.first
+        i += 1
+      end
+    else
+      count_array = self
+    end
+    count_array.length
+  end
```
`my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
the count can also perform with range too. But your one cannot perform with range. While I was playing with your method just came up with a formation that works. But need a lot of refactoring. Or you can generate your own form of solution. The issue I am facing while executing my_count with range
```ruby
(1...3).my_count # should return 2
(1...3).my_count(2) # should return 1
```
Now the solution that I have generated by playing with your code.
```ruby
def my_count(*args)
    arr = is_a?(Array) ? self : to_a
    count = 0
    i = 0
    if block_given?
      while i < arr.length
        count += 1 if yield arr[i]
        i += 1
      end
    elsif !block_given? && !args.empty?
      while i < arr.length
        count += 1 if arr[i] === args.first
        i += 1
      end
    else
      while i < arr.length
        count += 1 if arr[i]
        i += 1
      end
    end
    count
  end
```

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+    count_array.length
+  end
+
+  def my_map
+    object = to_a
+    modified_object = []
+
+    i = 0
+    if block_given?
+      while i < to_a.length
+        modified_object << (yield object[i])
+        i += 1
+      end
+      modified_object
+    else
+      to_enum
+    end
+  end
```
`my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
```ruby
my_proc = Proc.new { |num| num > 10 }
[11, 2, 3, 15].my_map(my_proc) {|num| num < 10 } # should return [true, false, false, true]
```
![image](https://user-images.githubusercontent.com/38625956/84071501-a04e8600-a9ef-11ea-88c2-e74e1b0c15e4.png)

shshamim63 commented on my_enumerables.rb:
```diff
@@ -0,0 +1,203 @@
+    end
+  end
+
+  def my_inject(initial = nil, method = nil)
+    modified_object = []
+    if block_given?
+      if initial
+        memo = initial
+        modified_object = my_map { |element| memo = yield memo, element }
+      else
+        memo = first
+        modified_object = to_a[1..-1].my_map { |element| memo = yield memo, element }
+      end
+    elsif !block_given?
+      if initial && method
+        memo = initial
+        modified_object = my_map { |element| memo = memo.send(method, element) }
+      elsif initial
+        method = initial
+        memo = first
+        modified_object = to_a[1..-1].my_map { |element| memo = memo.send(method, element) }
+      end
+    end
+    modified_object[-1]
+  end
```
my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- raises a "LocalJumpError" when no block or argument is given. But your one does not cover any case similar to this issue.

shshamim63 commented on .rubocop.yml:
```diff
@@ -0,0 +1,59 @@
+Layout/LineLength:
+  Max: 120
+Metrics/MethodLength:
+  Max: 40
```
[Optional]
You should not make a change to the config file. As the previous reviewer did not mention that I am marking it as optional. But remember, you should not make any change to this file.

shshamim63 reviewed (changes requested):
## Status :dart: Changes Required :dart: 
___
Hi @anewman15 :wave: 
Awesome job Almost covered all the changes asked by the previous reviewer. You just need to fix couple of corner cases to get approval. Let's fix them.
___
### :spades: Mandatory Changes :spades: 
___
- You were asked to make enumerable methods that are functional for array and hash classes(range, hash)
![image](https://user-images.githubusercontent.com/38625956/84072646-549cdc00-a9f1-11ea-9e6b-cdb720f7ec2e.png)
But your methods work for the array only. Please check the comments attached to the code block to resolve the issue.
___
### :clubs: Optional Changes :clubs: 
___
- You should not make any changes to the config files. in you have to disable any rules please do it using inline disabled.
___
##### Please fix the above-mentioned issues and send another review request. Feel free to knock me on slack for any explanation.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

anewman15 committed:
Make these changes according to TSE's suggestions:
- Convert self into array in my_each method
- Convert self into array in my_each_with_index method
- Move multiply_els outside Enumerable module method
- Convert self into array in my_select method
- Convert self into array in my_all? method

anewman15 committed:
The following changes:
- Convert self into array in my_any? method
- Convert self into array in my_none? method
- Replace return object with counter for my_count method
- Change my_map method to use only proc when both proc and block is given
- Modify my_inject method to output LocalJumpError when no block or argument is given
- Restore .rubocop.yml file and disable rubocop properties from my_enumerables.rb file

Forison reviewed (changes requested):
## Status:: Changes Required❌
# 🙋🏾‍♂️ Abdullah,
#### You have done a good job so far on addressing almost all the issues raised in the previous review, here are a few issues to address to get this project approved.

* [ ] Kindly make sure all the linter checks are passing before you submit the project for review. Note: linter are a mandatory requirement for every project. 

* [ ] Kindly address the issue raised earlier in relation to ```my_map``` thus
```my_proc = Proc.new { |num| num > 10 }```
```p [11, 2, 3, 15].my_map(my_proc) {|num| num < 10 }```
 ```# should return [true, false, false, true]```
### That is it, kindly make the changes and submit the project for re-review🧑‍✈️
#### Happy coding👨🏾‍💻
Addo Forison

anewman15 committed:
Combine my_map_proc_block method into my_map

anewman15 committed:
Comment out rubocop:enable... line and fix lint errors

Torres-ssf reviewed (approved):
### Project Approved :tada: 
Hi there @anewman15 
Great job on this project :+1: 
All required changes were implemented
You can merge this PR now

Happy coding :computer: 
**Sergio Torres**

anewman15 commented:
Thanks @lekegitrepo, @shshamim63 , @Forison and @Torres-ssf for your valuable feedback
